Attempt at writing guidance of which inputs types to use when in modules (#1486)
nf-core/nf-co.re: markdown/developers/modules.md

## Guidance on module inputs

There are various ways information that a given software needs can be entered into a module to make the module work.
We aim to make nf-core modules as flexible as possible, to ensure pipelines are not constrained by 'particularities' of a given module. In otherways, we want to make modules work for pipelines, not pipelines to work for modules.
Here we will provide some general guidance on how and when to use the various ways of passing information into a module.

The five ways of inserting information into a module are:

- `input:`: a distinct input channel defined by in the module itself
- `val(meta)`: a common entry in nf-core module tuple-based input channels that is a list of information associated with a given file
- `ext.args`: defined in `modules.conf` configuration files that is used to insert (normally non-mandatory) custom flags/options to the tools used in the module
- `ext.prefix`: defined in `modules.conf` configuration files that is used for defining file names (typically based on a `meta.id` element)
- `ext.<custom>`: defined in `modules.conf, custom 'variables' that can be passed into the module, defined by the writer of the module

> In general the priority order is:
> 
> - All files should be inserted into a module via `input:` channels
> - All non-file inputs: `args` > `meta` > `ext.custom`
>   - Where custom pipeline-specific `meta` keys can be inserted into an `args`, rather than module-hardcoded `meta` keys

Reviewer: ```suggestion
In general:

- All files, mandatory, or non-mandatory, must have a corresponding `input:` channel ( type `path` ).
- Command line options essential to the functioning of the tool, must also be `input:` channels ( type `val` ), or encoded as a sensible default that can be overridden by checking `ext.args`. 
- Filenames should be defined by `ext.prefix` and a file extension. 
- Sample specific command-line options should be defined based on fields of the `meta` map. If the command-line option is essential to the function of the tool, then the channel operator `multiMap` should be used to construct the string and pass it to an `input:` channel. If the command-line option is not essential to the operation of the tool or certain options should be defined by what is in the `meta` map, then the command-line options should be defined using `ext.args`.  
```

## ext.args

This is the main way to pass non-file inputs (flags, options, strings, etc.) to a tool within a module. Every nf-core module must include an `$args` variable within the command of the tool itself.

Pipeline developers can insert any information stored in their `meta` maps to this variable via a `modules.config` (see below). Users can also add their own additional personal parameters or information to this via a custom `modules.conf` file (at their own risk).

This can be done via closures, and be dynamically constructued using conditions within lists.

```nextflow

> ext.args = {
>   [
>     "-a" ,
>     "-b ${meta.id}",
>     run_mode ? "-c" : ""
>     ].join(' ') }

Reviewer: ```suggestion
ext.args = {
  [
    "-a" ,
    "-b ${meta.id}",       // Set's an option value to be the meta.id
    run_mode ? "-c" : ""   // run_mode is a variable defined in the main.nf of the module
    ].join(' ') }
```

```

> It is important to note that it is problematic to have something in ` config` that keeps a _module_ working, as configs can be overwritten. Therefore if the _module_ (not the tool!) requires a particular parameter to execute such information can be passed via an `input:` channel (see below). However, a pipeline can make sure that the _tool_ keeps working (i.e., non-file arguments) via their own `modules.conf`.

Reviewer: > However, a pipeline can make sure that the _tool_ keeps working (i.e., non-file arguments) via their own `modules.conf`. 

This is not true. A custom config completely overrides `ext.args` and if an option is supplied, the onus ends up on the user to make sure the tool keeps functioning.

Author: Ah true!

> It is important to note that it is problematic to have something in ` config` that keeps a _module_ working, as configs can be overwritten. Therefore if the _module_ (not the tool!) requires a particular parameter to execute such information can be passed via an `input:` channel (see below). However, a pipeline can make sure that the _tool_ keeps working (i.e., non-file arguments) via their own `modules.conf`.

Reviewer: ```suggestion
It is important to note that it is problematic to have something in a `config` that keeps a _module_ working, as configs can be overwritten. Therefore if the _module_ (not the tool!) requires a particular parameter to execute such information must be passed via an `input:` channel (see below).
```

- [Set ext.prefix based on task inputs](https://github.com/nf-core/rnaseq/blob/e049f51f0214b2aef7624b9dd496a404a7c34d14/conf/modules.config#L297)
- [Set ext.args based on both parameters and task inputs](https://github.com/nf-core/rnaseq/blob/e049f51f0214b2aef7624b9dd496a404a7c34d14/conf/modules.config#L377-L381)

## Guidance on module inputs

There are various ways information that a given software needs can be entered into a module to make the module work.
nf-core modules are intended to be flexible as possible, and ensure pipelines are not constrained by 'particularities' of a given module.
In other words, we want to make modules work for pipelines, not pipelines to work for modules.
Here we provide some general guidance on how and when to use the various ways of passing information into a module.

The four ways of passing information into a module are:

> - `input:`: a distinct input channel defined by in the module itself

Author: ```suggestion
- `input:`: a distinct input channel defined in the module itself
```

In other words, we want to make modules work for pipelines, not pipelines to work for modules.
Here we provide some general guidance on how and when to use the various ways of passing information into a module.

The four ways of passing information into a module are:

- `input:`: a distinct input channel defined in the module itself
- `val(meta)`: a common entry in nf-core module tuple-based input channels that is a list of information associated with a given sample/entity or associated files ( e.g. `meta.single_end` describing the endedness of the sequence files associated with the sample).
- `ext.args`: defined in `modules.conf` configuration files that is used to insert non-mandatory custom flags/options to the tools used in the module.
- `ext.prefix`: defined in `modules.conf` configuration files that is used for defining file names (typically based on a `meta.id` element).

Note: Custom ext fields `ext.<custom>` are discouraged as most use-cases can be accomodated using the above techniques.

In general:

- All files, mandatory, or non-mandatory, MUST have a corresponding `input:` channel ( type `path` ).

> - Command-line options essential to the functioning of the tool MUST also be `input:` channels ( type `val` ), or encoded as a sensible default that can be overridden by checking `args` (e.g., seeding options for reproducibilty).

Reviewer: True.
However, now that I've started actually using this, I feel that this way goes towards better maintainability. For those that are asked to add new modules to workflows, it signals quicker what needs to be added, as there's no record of what's a mandatory non-file options otherwise. And as I say above, it makes it easier to change workflow options for users.

> - Command-line options essential to the functioning of the tool MUST also be `input:` channels ( type `val` ), or encoded as a sensible default that can be overridden by checking `args` (e.g., seeding options for reproducibilty).

Reviewer 2: It could be a bit frustrating if we change the rules every couple of months tbh :D  we had a poll about it and all.

> - Command-line options essential to the functioning of the tool MUST also be `input:` channels ( type `val` ), or encoded as a sensible default that can be overridden by checking `args` (e.g., seeding options for reproducibilty).

Reviewer: I'm hoping this won't need to be changed again. Do we have good reasons for maintaining mandatory non-file options via `ext.args`?

> - Command-line options essential to the functioning of the tool MUST also be `input:` channels ( type `val` ), or encoded as a sensible default that can be overridden by checking `args` (e.g., seeding options for reproducibilty).

Reviewer 2: I really don't see an advantage of passing the mandatory params all through the pipeline and subworkflows. Imo it will add a ton of boilerplate code and make it much harder to see what a module is actually run with. How would you handle arguments that are mandatory but sample specific, i.e. rely on meta map info, that is not standardised across pipelines? The whole extra args/overwritting args thing has been fairly common for us. So the ease of just providing an updated config to users (and having this even more simplified with what @maxulysse is working on here https://github.com/nf-core/sarek/pull/891). For the rare modules that need some weird exception like controlfreec, ascat,gsea we can always have custom solutions of course.

> - Command-line options essential to the functioning of the tool MUST also be `input:` channels ( type `val` ), or encoded as a sensible default that can be overridden by checking `args` (e.g., seeding options for reproducibilty).

Reviewer: > I really don't see an advantage of passing the mandatory params all through the pipeline and subworkflows. 

- Pipeline stability, when a user changes `ext.args`. A user can only change what is meant to be configured by the user. The rest is kept in line with the workflow code.
- A module user can see explicitly which parameters are needed ( and clarified by the meta.yml ). 
  - Pipeline authors need not be as experienced with the tool to get it to work. 
- Single place in code where mandatory arguments can be checked, compared to both workflow and `modules.config`.

> Imo it will add a ton of boilerplate code and make it much harder to see what a module is actually run with. 

These channels are mandatory input. It should be less difficult, rather than thinking all the input is provided, and then it breaks because something is missing from `ext.args` (I've done this many times). Why would not being explicit about something make it harder to run? Also having all the mandatory input come through a single point of entry (i.e. workflow only) makes it easier to see what the module is taking as input. 

> How would you handle arguments that are mandatory but sample specific, i.e. rely on meta map info, that is not standardised across pipelines? 

This is explicit in the text. What can be done to make it clearer?

> The whole extra args/overwritting args thing has been fairly common for us. So the ease of just providing an updated config to users (and having this even more simplified with what @maxulysse is working on here [nf-core/sarek#891](https://github.com/nf-core/sarek/pull/891)). For the rare modules that need some weird exception like controlfreec, ascat,gsea we can always have custom solutions of course.

I think this proposed solution is not the right way to go about it. However, I'm not familiar enough with those modules to really understand what your needs are. The solution proposed keeps mandatory options explicit in separate places, and adds another option on top of the current way of providing input. While it may make the code more lean ( which I'm generally in favour of ), I think this does not really help readability, or understanding. I'm wondering if this is going in the direction of mandatory options for the pipeline, but not the module. In which case, I think that the module should have a patch file in the pipeline where the module code maintains stability for the extra parameters needed for the pipeline to remain functional in the face of changes to `ext.args`.

In my mind, `ext.args` are meant to be User configurable options, that don't break the workflow if changed. The onus should be on the pipeline developer to keep it stable, and that may need to be done using a patch.
